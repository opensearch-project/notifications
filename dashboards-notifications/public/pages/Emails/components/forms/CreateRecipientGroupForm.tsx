/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiComboBox,
  EuiComboBoxOptionOption,
  EuiFieldText,
  EuiFormRow,
  EuiSpacer,
  EuiText,
  EuiTextArea,
} from '@elastic/eui';
import React from 'react';
import { onComboBoxCreateOption } from '../../utils/helper';
import {
  validateRecipientGroupEmails,
  validateRecipientGroupName,
} from '../../utils/validationHelper';

interface CreateRecipientGroupFormProps {
  name: string;
  setName: (name: string) => void;
  description: string;
  setDescription: (name: string) => void;
  selectedEmailOptions: Array<EuiComboBoxOptionOption<string>>;
  setSelectedEmailOptions: (
    options: Array<EuiComboBoxOptionOption<string>>
  ) => void;
  emailOptions: Array<EuiComboBoxOptionOption<string>>;
  setEmailOptions: (options: Array<EuiComboBoxOptionOption<string>>) => void;
  inputErrors: { [key: string]: string[] };
  setInputErrors: (errors: { [key: string]: string[] }) => void;
}

export function CreateRecipientGroupForm(props: CreateRecipientGroupFormProps) {
  return (
    <>
      <EuiFormRow
        label="Name"
        style={{ maxWidth: '650px' }}
        helpText="The name must contain 2 to 50 characters. Valid characters are lowercase A-Z, a-z, 0-9, (_) underscore, (-) hyphen and unicode characters."
        error={props.inputErrors.name.join(' ')}
        isInvalid={props.inputErrors.name.length > 0}
      >
        <EuiFieldText
          fullWidth
          data-test-subj="create-recipient-group-form-name-input"
          placeholder="Enter recipient group name"
          value={props.name}
          onChange={(e) => props.setName(e.target.value)}
          isInvalid={props.inputErrors.name.length > 0}
          onBlur={() => {
            props.setInputErrors({
              ...props.inputErrors,
              name: validateRecipientGroupName(props.name),
            });
          }}
        />
      </EuiFormRow>

      <EuiSpacer size="m" />
      <EuiFormRow
        label={
          <span>
            Description - <i style={{ fontWeight: 'normal' }}>optional</i>
          </span>
        }
        style={{ maxWidth: '650px' }}
      >
        <>
          <EuiText size="xs" color="subdued">
            Describe the purpose of the channel.
          </EuiText>
          <EuiTextArea
            fullWidth
            data-test-subj="create-recipient-group-form-description-input"
            placeholder="What is the purpose of this recipient group?"
            style={{ height: '4.1rem' }}
            value={props.description}
            onChange={(e) => props.setDescription(e.target.value)}
          />
        </>
      </EuiFormRow>

      <EuiSpacer size="m" />
      <EuiFormRow
        label="Emails"
        style={{ maxWidth: '650px' }}
        error={props.inputErrors.emailOptions.join(' ')}
        isInvalid={props.inputErrors.emailOptions.length > 0}
      >
        <>
          <EuiText size="xs" color="subdued">
            Select or type in one or more email addresses.
          </EuiText>
          <EuiComboBox
            placeholder="Email addresses"
            data-test-subj="create-recipient-group-form-emails-input"
            fullWidth
            options={props.emailOptions}
            selectedOptions={props.selectedEmailOptions}
            onChange={props.setSelectedEmailOptions}
            onCreateOption={(searchValue, flattenedOptions) =>
              onComboBoxCreateOption(
                searchValue,
                flattenedOptions,
                props.emailOptions,
                props.setEmailOptions,
                props.selectedEmailOptions,
                props.setSelectedEmailOptions,
                (options) =>
                  props.setInputErrors({
                    ...props.inputErrors,
                    emailOptions: validateRecipientGroupEmails(options),
                  })
              )
            }
            customOptionText={'Add {searchValue} as an email address'}
            isClearable={true}
            isInvalid={props.inputErrors.emailOptions.length > 0}
            onBlur={() => {
              props.setInputErrors({
                ...props.inputErrors,
                emailOptions: validateRecipientGroupEmails(
                  props.selectedEmailOptions
                ),
              });
            }}
          />
        </>
      </EuiFormRow>

      <EuiSpacer size="m" />
    </>
  );
}
