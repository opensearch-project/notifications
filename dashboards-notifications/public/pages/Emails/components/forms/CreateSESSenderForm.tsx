/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiFieldText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiSpacer,
} from '@elastic/eui';
import React, { useContext } from 'react';
import { MainContext } from '../../../Main/Main';
import {
  validateAwsRegion,
  validateEmail,
  validateRoleArn,
  validateSenderName,
} from '../../utils/validationHelper';

interface CreateSESSenderFormProps {
  senderName: string;
  setSenderName: (name: string) => void;
  email: string;
  setEmail: (email: string) => void;
  roleArn: string;
  setRoleArn: (roleArn: string) => void;
  awsRegion: string;
  setAwsRegion: (awsRegion: string) => void;
  inputErrors: { [key: string]: string[] };
  setInputErrors: (errors: { [key: string]: string[] }) => void;
}

export function CreateSESSenderForm(props: CreateSESSenderFormProps) {
  const mainStateContext = useContext(MainContext)!;
  return (
    <>
      <EuiFormRow
        label="Sender name"
        style={{ maxWidth: '650px' }}
        helpText="Use a unique, descriptive name. The sender name must contain from 2 to 50 characters. Valid characters are lowercase a-z, 0-9, - (hyphen) and _ (underscore)."
        error={props.inputErrors.senderName.join(' ')}
        isInvalid={props.inputErrors.senderName.length > 0}
      >
        <EuiFieldText
          fullWidth
          placeholder="Enter sender name"
          data-test-subj="create-ses-sender-form-name-input"
          value={props.senderName}
          onChange={(e) => props.setSenderName(e.target.value)}
          isInvalid={props.inputErrors.senderName.length > 0}
          onBlur={() => {
            props.setInputErrors({
              ...props.inputErrors,
              senderName: validateSenderName(props.senderName),
            });
          }}
        />
      </EuiFormRow>

      <EuiSpacer size="m" />
      <EuiFormRow
        label="Email address"
        style={{ maxWidth: '650px' }}
        error={props.inputErrors.email.join(' ')}
        isInvalid={props.inputErrors.email.length > 0}
      >
        <EuiFieldText
          fullWidth
          placeholder="name@example.com"
          data-test-subj="create-ses-sender-form-email-input"
          value={props.email}
          onChange={(e) => props.setEmail(e.target.value)}
          isInvalid={props.inputErrors.email.length > 0}
          onBlur={() => {
            props.setInputErrors({
              ...props.inputErrors,
              email: validateEmail(props.email),
            });
          }}
        />
      </EuiFormRow>

      <EuiSpacer size="m" />
      <EuiFlexGroup gutterSize="s" style={{ maxWidth: '658px' }}>
        <EuiFlexItem grow={6}>
          <EuiFormRow
            label={
              mainStateContext.tooltipSupport ? (
                <span>
                  IAM Role ARN -{' '}
                  <i style={{ fontWeight: 'normal' }}>optional</i>
                </span>
              ) : (
                'IAM Role ARN'
              )
            }
            error={props.inputErrors.roleArn.join(' ')}
            isInvalid={props.inputErrors.roleArn.length > 0}
          >
            <EuiFieldText
              placeholder="ARN key"
              data-test-subj="create-ses-sender-form-role-arn-input"
              value={props.roleArn}
              onChange={(e) => props.setRoleArn(e.target.value)}
              isInvalid={props.inputErrors.roleArn.length > 0}
              onBlur={() => {
                if (!mainStateContext.tooltipSupport) {
                  props.setInputErrors({
                    ...props.inputErrors,
                    roleArn: validateRoleArn(props.roleArn),
                  });
                }
              }}
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem grow={4}>
          <EuiFormRow
            label="AWS region"
            error={props.inputErrors.awsRegion.join(' ')}
            isInvalid={props.inputErrors.awsRegion.length > 0}
          >
            <EuiFieldText
              placeholder="us-east-1"
              data-test-subj="create-ses-sender-form-aws-region-input"
              value={props.awsRegion}
              onChange={(e) => props.setAwsRegion(e.target.value)}
              isInvalid={props.inputErrors.awsRegion.length > 0}
              onBlur={() => {
                props.setInputErrors({
                  ...props.inputErrors,
                  awsRegion: validateAwsRegion(props.awsRegion),
                });
              }}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>

      <EuiSpacer size="m" />
    </>
  );
}
