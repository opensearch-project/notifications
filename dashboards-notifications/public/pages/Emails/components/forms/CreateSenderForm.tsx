/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiFieldNumber,
  EuiFieldText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiLink,
  EuiSpacer,
  EuiSuperSelect,
  EuiSuperSelectOption,
} from '@elastic/eui';
import React from 'react';
import {
  ALERTING_DOCUMENTATION_LINK,
  ENCRYPTION_TYPE,
} from '../../../../utils/constants';
import {
  validateEmail,
  validateHost,
  validatePort,
  validateSenderName,
} from '../../utils/validationHelper';

interface CreateSenderFormProps {
  senderName: string;
  setSenderName: (name: string) => void;
  email: string;
  setEmail: (email: string) => void;
  host: string;
  setHost: (host: string) => void;
  port: string;
  setPort: (port: string) => void;
  encryption: keyof typeof ENCRYPTION_TYPE;
  setEncryption: (encryption: keyof typeof ENCRYPTION_TYPE) => void;
  inputErrors: { [key: string]: string[] };
  setInputErrors: (errors: { [key: string]: string[] }) => void;
}

export function CreateSenderForm(props: CreateSenderFormProps) {
  const encryptionOptions: Array<EuiSuperSelectOption<
    keyof typeof ENCRYPTION_TYPE
  >> = Object.entries(ENCRYPTION_TYPE).map(([key, value]) => ({
    value: key as keyof typeof ENCRYPTION_TYPE,
    inputDisplay: value,
  }));

  return (
    <>
      <EuiFormRow
        label="Sender name"
        style={{ maxWidth: '650px' }}
        helpText="Use a unique, descriptive name. The sender name must contain from 2 to 50 characters. Valid characters are lowercase a-z, 0-9, - (hyphen) and _ (underscore)."
        error={props.inputErrors.senderName.join(' ')}
        isInvalid={props.inputErrors.senderName.length > 0}
      >
        <EuiFieldText
          fullWidth
          placeholder="Enter sender name"
          data-test-subj="create-sender-form-name-input"
          value={props.senderName}
          onChange={(e) => props.setSenderName(e.target.value)}
          isInvalid={props.inputErrors.senderName.length > 0}
          onBlur={() => {
            props.setInputErrors({
              ...props.inputErrors,
              senderName: validateSenderName(props.senderName),
            });
          }}
        />
      </EuiFormRow>

      <EuiSpacer size="m" />
      <EuiFlexGroup gutterSize="s" style={{ maxWidth: '658px' }}>
        <EuiFlexItem grow={4}>
          <EuiFormRow
            label="Email address"
            error={props.inputErrors.email.join(' ')}
            isInvalid={props.inputErrors.email.length > 0}
          >
            <EuiFieldText
              placeholder="name@example.com"
              data-test-subj="create-sender-form-email-input"
              value={props.email}
              onChange={(e) => props.setEmail(e.target.value)}
              isInvalid={props.inputErrors.email.length > 0}
              onBlur={() => {
                props.setInputErrors({
                  ...props.inputErrors,
                  email: validateEmail(props.email),
                });
              }}
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem grow={4}>
          <EuiFormRow
            label="Host"
            error={props.inputErrors.host.join(' ')}
            isInvalid={props.inputErrors.host.length > 0}
          >
            <EuiFieldText
              placeholder="smtp.example.com"
              data-test-subj="create-sender-form-host-input"
              value={props.host}
              onChange={(e) => props.setHost(e.target.value)}
              isInvalid={props.inputErrors.host.length > 0}
              onBlur={() => {
                props.setInputErrors({
                  ...props.inputErrors,
                  host: validateHost(props.host),
                });
              }}
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem grow={2}>
          <EuiFormRow
            label="Port"
            error={props.inputErrors.port.join(' ')}
            isInvalid={props.inputErrors.port.length > 0}
          >
            <EuiFieldNumber
              placeholder="465"
              data-test-subj="create-sender-form-port-input"
              value={props.port}
              onChange={(e) => props.setPort(e.target.value)}
              isInvalid={props.inputErrors.port.length > 0}
              onBlur={() => {
                props.setInputErrors({
                  ...props.inputErrors,
                  port: validatePort(props.port),
                });
              }}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>

      <EuiSpacer size="m" />
      <EuiFormRow
        label="Encryption method"
        style={{ maxWidth: '650px' }}
        helpText={
          <div>
            SSL or TLS is recommended for security. To use either one, you must
            enter each sender account's credentials to the OpenSearch
            keystore using the CLI.{' '}
            <EuiLink
              href={ALERTING_DOCUMENTATION_LINK}
              target="_blank"
              external
            >
              Learn more
            </EuiLink>
          </div>
        }
      >
        <EuiSuperSelect
          fullWidth
          data-test-subj="create-sender-form-encryption-input"
          options={encryptionOptions}
          valueOfSelected={props.encryption}
          onChange={props.setEncryption}
        />
      </EuiFormRow>

      <EuiSpacer size="m" />
    </>
  );
}
