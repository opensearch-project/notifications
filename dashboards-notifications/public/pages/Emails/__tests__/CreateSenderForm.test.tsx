/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { render } from '@testing-library/react';
import React from 'react';
import { CreateSenderForm } from '../components/forms/CreateSenderForm';

describe('<CreateSenderForm/> spec', () => {
  it('renders the component', () => {
    const setSenderName = jest.fn();
    const setEmail = jest.fn();
    const setHost = jest.fn();
    const setPort = jest.fn();
    const setEncryption = jest.fn();
    const utils = render(
      <CreateSenderForm
          senderName="test"
          setSenderName={setSenderName}
          email="test"
          setEmail={setEmail}
          host="test"
          setHost={setHost}
          port="test"
          setPort={setPort}
          encryption="SSL"
          setEncryption={setEncryption}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });
});
