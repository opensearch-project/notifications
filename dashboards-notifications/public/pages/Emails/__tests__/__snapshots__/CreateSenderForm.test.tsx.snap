// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateSenderForm/> spec renders errors 1`] = `
<div
  class="euiFormRow"
  id="random_html_id-row"
  style="max-width: 650px;"
>
  <div
    class="euiFormRow__labelWrapper"
  >
    <label
      aria-invalid="true"
      class="euiFormLabel euiFormRow__label euiFormLabel-isInvalid"
      for="random_html_id"
    >
      Sender name
    </label>
  </div>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <div
      class="euiFormControlLayout euiFormControlLayout--fullWidth"
    >
      <div
        class="euiFormControlLayout__childrenWrapper"
      >
        <input
          aria-describedby="random_html_id-help-0 random_html_id-error-0"
          class="euiFieldText euiFieldText--fullWidth"
          data-test-subj="create-sender-form-name-input"
          id="random_html_id"
          placeholder="Enter sender name"
          type="text"
          value="test"
        />
      </div>
    </div>
    <div
      aria-live="polite"
      class="euiFormErrorText euiFormRow__text"
      id="random_html_id-error-0"
    >
      test error
    </div>
    <div
      class="euiFormHelpText euiFormRow__text"
      id="random_html_id-help-0"
    >
      Use a unique, descriptive name. The sender name must contain from 2 to 50 characters. Valid characters are lowercase a-z, 0-9, and - (hyphen).
    </div>
  </div>
</div>
`;

exports[`<CreateSenderForm/> spec renders the component 1`] = `
<div
  class="euiFormRow"
  id="random_html_id-row"
  style="max-width: 650px;"
>
  <div
    class="euiFormRow__labelWrapper"
  >
    <label
      aria-invalid="false"
      class="euiFormLabel euiFormRow__label"
      for="random_html_id"
    >
      Sender name
    </label>
  </div>
  <div
    class="euiFormRow__fieldWrapper"
  >
    <div
      class="euiFormControlLayout euiFormControlLayout--fullWidth"
    >
      <div
        class="euiFormControlLayout__childrenWrapper"
      >
        <input
          aria-describedby="random_html_id-help-0"
          class="euiFieldText euiFieldText--fullWidth"
          data-test-subj="create-sender-form-name-input"
          id="random_html_id"
          placeholder="Enter sender name"
          type="text"
          value="test"
        />
      </div>
    </div>
    <div
      class="euiFormHelpText euiFormRow__text"
      id="random_html_id-help-0"
    >
      Use a unique, descriptive name. The sender name must contain from 2 to 50 characters. Valid characters are lowercase a-z, 0-9, and - (hyphen).
    </div>
  </div>
</div>
`;
